### ISSUE TITLE
Customized mlogloss eval_metric in R

### ISSUE THREAD

**Issue by yib2** [dmlc/mxnet#763]
I tried to customize mlogloss evaluation metric, but when i set a small array batch size, it failed to match the length of label and pred. Does anybody try to write a customized mlogloss function in R?

**Comment by jermainewang** [comment]
@tqchen Could you help on this?

**Comment by tqchen** [comment]
@hetong007

**Comment by tqchen** [comment]
@thirdwing

**Comment by thirdwing** [comment]
@yib2 Can you share the function you used?

**Comment by yib2** [comment]
mx.metric.mlogloss <- mx.metric.custom("mlogloss", function(label, pred){
  
  label <- as.factor(label)
  label_mat <- matrix(0, nrow = length(label), ncol = length(levels(label)))
  sample_levels <- as.integer(label)
  for (i in 1:length(label)) label_mat[i, sample_levels[i]] <- 1
  label <- label_mat
  #I tried to multiply label sparse matrix with prediction matrix, but it returned an error, like feval() something. 

  label <- as.vector(t(label))
  eps <- 1e-15
  n <- nrow(pred)
  pred <- pmax(pmin(pred, 1 - eps), eps)
  MultiLogLoss <- (-1/n) * sum(label * log(pred))
  return(MultiLogLoss)
})

This is basically from package MLmetrics. It works well for mnist example, but the mlogloss may change while tuning the batch size. I think It only calculates a small proportion of data.

**Comment by thirdwing** [comment]
@yib2  It will be appreciated if you can also provide the code caused errors.

**Comment by yib2** [comment]
@thirdwing 

setwd('/Users/yib2/Documents/Graduate/Kaggle/Digit_Recognizer')
require(mxnet)

mx.metric.mlogloss <- mx.metric.custom("mlogloss", function(label, pred){
  label <- as.factor(label)
  label_mat <- matrix(0, nrow = length(label), ncol = length(levels(label)))
  sample_levels <- as.integer(label)
  for (i in 1:length(label)) label_mat[i, sample_levels[i]] <- 1
  label <- label_mat

  label <- as.vector(t(label))
  eps <- 1e-15
  n <- nrow(pred)
  pred <- pmax(pmin(pred, 1 - eps), eps)
  MultiLogLoss <- (-1/n) * sum(label * log(pred))
  return(MultiLogLoss)
})

train <- read.csv('train.csv', header=TRUE)
test <- read.csv('test.csv', header=TRUE)
train <- data.matrix(train)
test <- data.matrix(test)

train.x <- train[,-1]
train.y <- train[,1]

train.x <- t(train.x/255)
test <- t(test/255)

data <- mx.symbol.Variable("data")
fc1 <- mx.symbol.FullyConnected(data, name="fc1", num_hidden=128)
act1 <- mx.symbol.Activation(fc1, name="relu1", act_type="tanh")
fc2 <- mx.symbol.FullyConnected(act1, name="fc2", num_hidden=64)
act2 <- mx.symbol.Activation(fc2, name="relu2", act_type="tanh")
fc3 <- mx.symbol.FullyConnected(act2, name="fc3", num_hidden=10)
softmax <- mx.symbol.SoftmaxOutput(fc3, name="sm")

devices <- mx.cpu()
mx.set.seed(0)
model <- mx.model.FeedForward.create(softmax, X=train.x, y=train.y,
                                     ctx=devices, num.round=100, array.batch.size=100,
                                     learning.rate=0.07, momentum=0.9,  eval.metric=mx.metric.accuracy,
                                     initializer=mx.init.uniform(0.07),
                                     epoch.end.callback=mx.callback.log.train.metric(100))

Error in feval(as.array(label), as.array(pred)) :    dims [product 1000] do not match the length of object [2700]
When I used it on mnist example, it worked well. But this error raised when I used it on my project.

**Comment by thirdwing** [comment]
So do you mind sharing the data?

Without the data to reproduce the error, I can only guess the error is from 
```r
label <- as.vector(t(label))
``` 

or
```
MultiLogLoss <- (-1/n) * sum(label * log(pred))
```

**Comment by yib2** [comment]
My data is big. You can change the array.batch.size in minst example, and you will see the length of label and pred is changing, so as for mlogloss itself. 

you may add a line: print(length(label)) to see how it happened.

**Comment by thirdwing** [comment]
I think I know what you mean here and I just want to double check with your data.

In 

```r
label_mat <- matrix(0, nrow = length(label), ncol = length(levels(label)))
```

You have made an assumption that `label` has labels for all classes, which is wrong when `array.batch.size` is small.

**Comment by yib2** [comment]
Yes. You got it. For a specific problem, I think i can fix it by setting the number of columns, but it is not suitable for general environment. That would be great if you add a new eval_metric in mxnet.
Thanks.

**Comment by thirdwing** [comment]
You can use the line below for now.

```r
label_mat <- matrix(0, nrow = length(label), ncol = length(pred) / length(label) )
```
